treewide: load structured attributes in all bash builders consistently

wolfgangwalther opened the pull request:
It's hard to put the sourcing of ./.attrs.sh into all builder consistently - mistakes will happen. Thus, load structured attrs once in make-derivation and then source the remaining builder on top.

This should fix quite a few builders with structured attributes in principle. Most importantly it helps substitute / substituteAll, which are required for bootstrap on some platforms.

Part of the overall effort to enable structuredAttrs by default eventually, tracking: https://github.com/NixOS/nixpkgs/issues/205690

As the author of #257919, you might be interested in reviewing this, @ma27?

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
treewide: load structured attributes in all bash builders consistently

It's hard to put the sourcing of ./.attrs.sh into all builder
consistently - mistakes will happen. Thus, load structured attrs once in
make-derivation and then source the remaining builder on top.

This should fix quite a few builders with structured attributes in
principle. Most importantly it helps substitute / substituteAll, which
are required for bootstrap on some platforms.

wolfgangwalther reviewed (commented):
All the files which already load the attrs file before sourcing stdenv are straight-forward.

A lot of files can easily be verified to be passed to `mkDerivation`'s `builder` argument. Those are the cases that are straight up broken with structuredAttrs and easily fixed by loading the attrs file globally.

There are some remaining cases which are more special and need more attention. I left comments on each of them. Maybe I need to split some of these into a separate clean-up commit.

emilazy commented on pkgs/by-name/xq/xquartz/system-fonts.nix:
```diff
@@ -6,8 +6,6 @@ stdenv.mkDerivation {
     xorg.mkfontdir xorg.mkfontscale
   ];
   buildCommand = ''
-    source $stdenv/setup
-
```
XQuartz used to build, it’s just a regression. Probably wouldn’t be too hard to get it working again, although I don’t know if anyone cares enough to.

wolfgangwalther committed:
dockerTools.streamLayeredImage: fix with structuredAttrs

When building with structuredAttrs, two things would happen:
- sourcing stdenv would fail, because $stdenv is not set.
- the error is swallowed because bash is called without -e.

Thus, the build appears to be successful at first, but could fail later
when anything depends on a sourced stdenv.

wolfgangwalther committed:
various: remove useless "source stdenv/setup"

Those are run inside buildCommand, which is evaled inside stdenv/setup -
so no reason to source it again.

wolfgangwalther committed:
nixos/tasks: remove obsolete tty-backgrounds-combine.sh file

This should have been removed in
82e1d0c8b1b82345703caaf69afe94a809586e63.

wolfgangwalther committed:
nixosTests.etc: fix with structuredAttrs

The NIX_ATTRS_SH_FILE needs to be available visible inside the
fakechroot environment, so that we can load it for structuredAttrs
support.

wolfgangwalther reviewed (commented):
Found a few more cases of `source $stdenv/setup` outside `pkgs/`. Added fixes for the simple ones, but still need to look into the last two commits I added.

FTR: I can't make any sense of this at all, yet, will probably leave it out of this PR:

https://github.com/NixOS/nixpkgs/blob/d225878f0f482617b5388450932ade6ec6927149/pkgs/development/web/nodejs/node-npm-build-npm-package-logic.patch#L53

Putting in on draft until I looked at the last two commits again.

Ma27 commented:
> As the author of #257919, you might be interested in reviewing this, @Ma27?

Oh absolutely, can't guarantee that I'll get to it tomorrow, though!

Ma27 commented on pkgs/build-support/docker/default.nix:
```diff
@@ -969,7 +969,8 @@ rec {
             ${if enableFakechroot then ''
-              proot -r $PWD/old_out ${bind-paths} --pwd=/ fakeroot bash -c '
+              proot -r $PWD/old_out ${bind-paths} --pwd=/ fakeroot bash -e -c '
+                if [ -e "$NIX_ATTRS_SH_FILE" ]; then . "$NIX_ATTRS_SH_FILE"; elif [ -f .attrs.sh ]; then . .attrs.sh; fi
```
To give a bit of context: `NIX_ATTRS_*_FILE` was introduced by me in 2021 because writing attrs into `NIX_BUILD_TOP` was a little awkward for `nix-shell` use-cases where two `nix-shell` invocations in the same directory would discard each other's `.attrs.*` files. The idea was to be able to write those files into a different locations.

However, this is not the case on Nix 2.3 (I just realized that this wasn't backported, so 2.3.17 still doesn't have the relevant patch).

Now I see the following options

* bump minver to 2.4+ (I doubt this will happen)
* backport to 2.3.18 and bump minver once again
* keep the .attrs.* hack everywhere (I think we also need to add it to `runInLinuxVM` I just realized).

Damn, I should've caught the last aspect in the runInLinuxVM aspect. Perhaps I should take a break of a few days from reviewing code here :melting_face:

Ma27 commented on nixos/modules/system/etc/test.nix:
```diff
@@ -64,7 +64,14 @@ lib.recurseIntoAttrs {
+        if [ -e "$NIX_ATTRS_SH_FILE" ]; then . "$NIX_ATTRS_SH_FILE"; fi
+        source $stdenv/setup
+        eval "$fakeRootCommands"
+      '
```
This is being built with the local Nix, so it'd fail with 2.3, correct?

Ma27 reviewed (commented):
The change is a very good idea and I like it.
I can run some tests tomorrow on it and see if I can backport the NIX_ATTRS_*_FILE stuff in a minimal fashion to 2.3.

Until then, we should fix up the two remaining places where we only check for NIX_ATTRS_SH_FILE I'd argue.

Does this make sense?

Ma27 commented:
@wolfgangwalther https://github.com/NixOS/nix/pull/12026

Ma27 commented:
So I successfully[1] built a bunch of derivations with __structuredAttrs and confirmed that the changes to the "custom" builders are fine by testing randomly selected builders. As I said, I think we'd either need to fix the part I mentioned above and `runInLinuxVM` or await the backport I made and bump minver.

Other than that, the change looks pretty good & reasonable, thanks!

[1] Well, glibc-locales ffom the stdenv bootstrap fails with

<details><summary>Build error for `/nix/store/nl1lb52i77bhvq0yyahwd4j100phx0fv-glibc-locales-2.40-36.drv`</summary>

```
mv -f /build/build/Versions.mkT /build/build/Versions.mk
printf "#include <time64-compat.h>\n#ifdef TIME64_NON_DEFAULT\nhave-time64-compat = yes\n#endif" \
| gcc -E -undef    -Iinclude   -I/build/build  -Isysdeps/unix/sysv/linux/x86_64/64  -Isysdeps/unix/sysv/linux/x86_64/include -Isysdeps/unix/sysv/linux/x86_64  -Isysdeps/unix/sysv/linux/x86/include -Isysdeps/unix/sysv/linux/x86  -Isysdeps/x86/nptl  -Isysdeps/unix/sysv/linux/wordsize-64  -Isysdeps/x86_64/nptl  -Isysdeps/unix/sysv/linux/include -Isysdeps/unix/sysv/linux  -Isysdeps/nptl  -Isysdeps/pthread  -Isysdeps/gnu  -Isysdeps/unix/inet  -Isysdeps/unix/sysv  -Isysdeps/unix/x86_64  -Isysdeps/unix  -Isysdeps/posix  -Isysdeps/x86_64/64  -Isysdeps/x86_64/fpu/multiarch  -Isysdeps/x86_64/fpu  -Isysdeps/x86/fpu  -Isysdeps/x86_64/multiarch  -Isysdeps/x86_64  -Isysdeps/x86/include -Isysdeps/x86  -Isysdeps/ieee754/float128  -Isysdeps/ieee754/ldbl-96/include -Isysdeps/ieee754/ldbl-96  -Isysdeps/ieee754/dbl-64  -Isysdeps/ieee754/flt-32  -Isysdeps/wordsize-64  -Isysdeps/ieee754  -Isysdeps/generic   -Ilibio -I. -nostdinc -isystem /nix/store/y6ylhhrkn31fd0sgbx0cwrmbbfijf5ls-gcc-14-20241116/lib/gcc/x86_64-unknown-linux-gnu/14.2.1/include -isystem /nix/store/y6ylhhrkn31fd0sgbx0cwrmbbfijf5ls-gcc-14-20241116/lib/gcc/x86_64-unknown-linux-gnu/14.2.1/include-fixed -isystem no -D_LIBC_REENTRANT  -include include/libc-symbols.h       -DTOP_NAMESPACE=glibc -x assembler-with-cpp - > /build/build/time64-compat.mk.iT
mv -f /build/build/time64-compat.mk.iT /build/build/time64-compat.mk.i
sed '/^[        ]*#/d;/^[       ]*$/d' /build/build/time64-compat.mk.i > /build/build/time64-compat.mkT
mv -f /build/build/time64-compat.mkT /build/build/time64-compat.mk
gawk -v buildlist="iconvprogs iconvdata ldconfig libmemusage libpcprofile librpcsvc locale_programs memusagestat nonlib nscd extramodules libnldbl libsupport testsuite testsuite_internal" -f scripts/gen-libc-modules.awk /build/build/soversions.i > /build/build/libc-modules.T
/nix/store/razasrvdg7ckplfmvdxv4ia3wbayr94s-bootstrap-tools/bin/bash scripts/move-if-change /build/build/libc-modules.T /build/build/libc-modules.h
touch /build/build/libc-modules.stmp
for dir in /build/build sysdeps/unix/sysv/linux/x86_64/64 sysdeps/unix/sysv/linux/x86_64 sysdeps/unix/sysv/linux/x86 sysdeps/x86/nptl sysdeps/unix/sysv/linux/wordsize-64 sysdeps/x86_64/nptl sysdeps/unix/sysv/linux sysdeps/nptl sysdeps/pthread sysdeps/gnu sysdeps/unix/inet sysdeps/unix/sysv sysdeps/unix/x86_64 sysdeps/unix sysdeps/posix sysdeps/x86_64/64 sysdeps/x86_64/fpu/multiarch sysdeps/x86_64/fpu sysdeps/x86/fpu sysdeps/x86_64/multiarch sysdeps/x86_64 sysdeps/x86 sysdeps/ieee754/float128 sysdeps/ieee754/ldbl-96 sysdeps/ieee754/dbl-64 sysdeps/ieee754/flt-32 sysdeps/wordsize-64 sysdeps/ieee754 sysdeps/generic; do \
  test -f $dir/syscalls.list && \
  { sysdirs='sysdeps/unix/sysv/linux/x86_64/64 sysdeps/unix/sysv/linux/x86_64 sysdeps/unix/sysv/linux/x86 sysdeps/x86/nptl sysdeps/unix/sysv/linux/wordsize-64 sysdeps/x86_64/nptl sysdeps/unix/sysv/linux sysdeps/nptl sysdeps/pthread sysdeps/gnu sysdeps/unix/inet sysdeps/unix/sysv sysdeps/unix/x86_64 sysdeps/unix sysdeps/posix sysdeps/x86_64/64 sysdeps/x86_64/fpu/multiarch sysdeps/x86_64/fpu sysdeps/x86/fpu sysdeps/x86_64/multiarch sysdeps/x86_64 sysdeps/x86 sysdeps/ieee754/float128 sysdeps/ieee754/ldbl-96 sysdeps/ieee754/dbl-64 sysdeps/ieee754/flt-32 sysdeps/wordsize-64 sysdeps/ieee754 sysdeps/generic' \
    asm_CPP='gcc -c    -Iinclude   -I/build/build  -Isysdeps/unix/sysv/linux/x86_64/64  -Isysdeps/unix/sysv/linux/x86_64/include -Isysdeps/unix/sysv/linux/x86_64  -Isysdeps/unix/sysv/linux/x86/include -Isysdeps/unix/sysv/linux/x86  -Isysdeps/x86/nptl  -Isysdeps/unix/sysv/linux/wordsize-64  -Isysdeps/x86_64/nptl  -Isysdeps/unix/sysv/linux/include -Isysdeps/unix/sysv/linux  -Isysdeps/nptl  -Isysdeps/pthread  -Isysdeps/gnu  -Isysdeps/unix/inet  -Isysdeps/unix/sysv  -Isysdeps/unix/x86_64  -Isysdeps/unix  -Isysdeps/posix  -Isysdeps/x86_64/64  -Isysdeps/x86_64/fpu/multiarch  -Isysdeps/x86_64/fpu  -Isysdeps/x86/fpu  -Isysdeps/x86_64/multiarch  -Isysdeps/x86_64  -Isysdeps/x86/include -Isysdeps/x86  -Isysdeps/ieee754/float128  -Isysdeps/ieee754/ldbl-96/include -Isysdeps/ieee754/ldbl-96  -Isysdeps/ieee754/dbl-64  -Isysdeps/ieee754/flt-32  -Isysdeps/wordsize-64  -Isysdeps/ieee754  -Isysdeps/generic   -Ilibio -I. -nostdinc -isystem /nix/store/y6ylhhrkn31fd0sgbx0cwrmbbfijf5ls-gcc-14-20241116/lib/gcc/x86_64-unknown-linux-gnu/14.2.1/include -isystem /nix/store/y6ylhhrkn31fd0sgbx0cwrmbbfijf5ls-gcc-14-20241116/lib/gcc/x86_64-unknown-linux-gnu/14.2.1/include-fixed -isystem no -D_LIBC_REENTRANT -include /build/build/libc-modules.h -DMODULE_NAME=libc -include include/libc-symbols.h       -DTOP_NAMESPACE=glibc -DASSEMBLER  -g -Werror=undef -Wa,--noexecstack      -E -x assembler-with-cpp' \
    /nix/store/razasrvdg7ckplfmvdxv4ia3wbayr94s-bootstrap-tools/bin/bash sysdeps/unix/make-syscalls.sh $dir || exit 1; }; \
  test $dir = sysdeps/unix && break; \
done > /build/build/sysd-syscallsT
mv -f /build/build/sysd-syscallsT /build/build/sysd-syscalls
make  -C localedata install-locales
make[2]: Entering directory '/build/glibc-2.40/localedata'
Makefile:567: target 'install-archive-C.UTF-8/UTF-8' given more than once in the same rule
Makefile:571: target 'install-files-C.UTF-8/UTF-8' given more than once in the same rule
.././scripts/mkinstalldirs /nix/store/xlq8w798knyvsw1wwfnillggy18y76v2-glibc-locales-2.40-36/lib/locale
mkdir -p -- /nix/store/xlq8w798knyvsw1wwfnillggy18y76v2-glibc-locales-2.40-36/lib/locale
aa_DJ.UTF-8...cannot open locale archive "/nix/store/d5paaz91bbg7ycz3vz2c7mxznmb0zr1j-glibc-2.40-36/lib/locale/locale-archive": Permission denied
make[2]: *** [Makefile:568: install-archive-aa_DJ.UTF-8/UTF-8] Error 1
make[2]: Leaving directory '/build/glibc-2.40/localedata'
make[1]: *** [Makefile:760: localedata/install-locales] Error 2
make[1]: Leaving directory '/build/glibc-2.40'
make: *** [Makefile:9: localedata/install-locales] Error 2
/nix/store/2bd1233710hmlfkdyayrji87rzvc1kvq-bootstrap-stage4-stdenv-linux/setup: line 1759: pop_var_context: head of shell_variables not a function context
/nix/store/2bd1233710hmlfkdyayrji87rzvc1kvq-bootstrap-stage4-stdenv-linux/setup: line 15: pop_var_context: head of shell_variables not a function context
```

</details>

when doing `config.structuredAttrsByDefault = true;`, but this doesn't look like a new regression, so I fell back to testing individual packages with structured attrs enabled.

philiptaron commented:
> @wolfgangwalther [NixOS/nix#12026](https://github.com/NixOS/nix/pull/12026)

This is merged (thank you @edolstra!).

So is this the plan?

1. Tag 2.3.18 in the Nix repository
2. Open and merge a PR for Nix 2.3.18, likely updating `lib/minver.nix` at the same time.
3. Merge this PR

Let me know @Ma27 @wolfgangwalther

Ma27 commented:
> Does this mean we need to wait until minVersion > 2.15

Things that will happen before: flakes will be stabilized, I'll finish university.

> Am I missing anything?

No, I think you're right. It kinda sucks that this took so long to surface, but that's probably the price we have to pay given that this feature and everything around it was unused for so long and thus got almost no testing.

I see the following options:

* bump minver, but still allow 2.3 as an exception now that it behaves correctly.
* keep the `-f .attrs.sh` fallback (meh).

Minver is bumped very conservatively since you'd essentially block people from very old systems to upgrade in one go.
OTOH these versions are just dead and don't receive any support as opposed to Nix 2.3 and at some point you'll have to make a cut and now we have a compelling reason to do so.

I'm not convinced that we'll make the "minver"-call quickly, so my suggestion would be to continue with the `-f .attrs.sh` for now and defer the other part. That being said, I think it's still worth gathering opinions now to move forward.

---

cc @mweinelt @infinisil @K900 @emilazy 
To summarize the above: we'd like to be able to use `NIX_ATTRS_SH_FILE` ONLY for `__structuredAttrs = true;` which points to the location of `.attrs.sh` (which is not necessarily `NIX_BUILD_TOP`, e.g. for `nix develop`). This change is now in the 2.3 release branch of Nix. However, the feature is broken in Nix from 2.4 to 2.15 because it points to the wrong location in the build sandbox[1].
We'd like to get rid of the `if [ -f $NIX_ATTRS_SH_FILE ]; then ... elif [ -f .attrs.sh ]` stuff (not only in stdenv, but in any derivation using a custom builder). The idea I have is to bump minver aggressively (i.e. to 2.15), but still keep Nix 2.3 as an exception given somewhat popular demand (and the fact that it's still maintained).
This would mean that you'll need to switch to 23.11 first before you can upgrade to 25.05 since that one has 2.18 as default Nix. Upgrades fro 23.05 would still be possible if we allow the latest patch-versions of 2.13 and 2.14.

I decided to just ping a bunch of people who are heavily involved in nixpkgs given that we don't have a real architecture team that I could consult on that.

[1] seems as if it was my patch that introduced the issue, but it took almost two years until it got caught since `__structuredAttrs` is hardly used. But still, apologies for that!

infinisil commented:
@Ma27 I'm fine with bumping `minver.nix` to say 2.15 myself, but with the fallback it sounds like it's not needed anymore, which is even better :D

infinisil commented:
I see, let's try bumping it then. We should ping relevant parties (all Nix implementations) in the PR doing that and give plenty of time to react.

We could also consider somehow making the distinction between "version needed for eval" and "version needed for building", because we'll leave the former at 2.3, but the latter will be 2.15.

philiptaron commented:
> We could also consider somehow making the distinction between "version needed for eval" and "version needed for building", because we'll leave the former at 2.3, but the latter will be 2.15.

I'm pretty opposed to this split. It's not clear to me it provides any value. The Nix deprecation story is messy enough as it is. If we bump lib/minver.nix, it should be a thorough bump, I think.

For another thread/PR, though, I suppose.

Ma27 commented:
> The Nix deprecation story is messy enough as it is

I don't think this is strictly related to the Nix deprecation story: minver is supposed to be bumped conservatively to give even older instances a chance to upgrade.

The 2.3 thing just became a very hot topic since 2.4+ are somewhat controversial (tl;dr flakes and a very large diff) which is why I'm a little hesitant to write it off (also, it's still maintained).

I fully agree with what @wolfgangwalther said here, we should stop supporting Nix versions to get rid of workarounds like that (I mean, there are a bunch of places that don't check for both cases, so technically 2.3 is broken there currently) and I think this is the quicker way to achieve that. Also, I think Tvix is targeting Nix 2.3 as baseline?

Anyways, at the end of the day, somebody has to make that call. I think we even had a Nixpkgs architecture team in the past ;-)

> We could also consider somehow making the distinction between "version needed for eval" and "version needed for building"

I get the motivation, but wouldn't it be awkward if your Nix is sufficient to evaluate, but not to build nixpkgs?

Anyways, @wolfgangwalther the current state looks reasonable. I'd go through the changes soonish one last time and merge then.

Ma27 commented:
Meh this now needs a rebase @wolfgangwalther. Feel free to merge after that.

emilazy commented:
Late to the party here, but the Nix minimum version bump that makes sense to me is 2.18, as the “new 2.3” and a reasonable baseline for Nix semantics, given Lix’s explicit choice of it as a base and the year that Nixpkgs was stuck on it. That cuts off 23.05, but the upside is that 23.11 doesn’t just contain 2.18 but also *defaults* to it, so the story is pretty simple: “upgrade to 23.11 before upgrade to 25.05”.

I don’t think that anyone really cares about the versions between 2.4 and 2.18, and the latter has had an above‐average level of maintenance put into it given its use as the default version across multiple Nixpkgs releases. I know Tvix is aiming for 2.3 compatibility, but I believe that relates to post‐2.3 eval behaviour regressions that Nixpkgs assuming any newer version might run into anyway, and I imagine there is room for adjustment there given the importance of Nixpkgs compatibility. (But of course we should ping representatives of all implementations, and anyone else who might have a stake in the choice, on the PR.)

Ma27 commented:
I'm not really opposed to your suggestion, I just tried to be as conservative as possible to get it through. But I'm not sure I'll follow through now that things got a little more complicated.

emilazy commented:
Sorry if I contributed to any of the complication :) I do think some bump is a reasonable idea and I’m not too fussy about what exact version.

Ma27 commented:
> Sorry if I contributed to any of the complication :)

I think I used the wrong wording here, all good!
With this patch we have a single place that needs to cover the .attrs.sh vs. NIX_ATTRS_SH_FILE case (that I'm responsible for since I messed up the very first patch for that and it wasn't noticed for quite a while - hence there's a bunch of Nix minors being broken in that regard). For me, the current solution is good enough to not invest energy in discussing a big minver bump that I'd expect given past experiences of efforts trying to drop 2.3 (my personal expectation is that most opposition will come from 2.3 being dropped).

That being said, I do think your approach is reasonable and I'd be in favor of such a change.

emilazy commented:
Gotcha. Admittedly the (IMO premature) removal of 2.18 from Nixpkgs might make dropping 2.3 more controversial still. (Though probably most people who would move to that would just use Lix instead these days, I suppose?)

philiptaron commented:
Bisect says that this broke `nix-build -A tests.stdenv.outputs-no-out`. It fails with the same error as #214822 but is not that issue, I think, since @emilazy fixed it in August.

The only output is:

```
/nix/store/shkw4qm9qcw5sc5n1k5jznc83ny02r39-default-builder.sh: line 1: pop_var_context: head of shell_variables not a function context
```
